File: card-api/src/main/java/com/jean202/cardmizer/api/demo/RecommendationDemoScenariosLoader.java
package com.jean202.cardmizer.api.demo;

import java.io.IOException;
import java.io.InputStream;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import org.yaml.snakeyaml.Yaml;

public class RecommendationDemoScenariosLoader {
    static final String DEFAULT_RESOURCE_PATH = "demo/recommendation-scenarios.yml";

    public RecommendationDemoScenarios loadDefault() {
        return load(DEFAULT_RESOURCE_PATH);
    }

    public RecommendationDemoScenarios load(String resourcePath) {
        Objects.requireNonNull(resourcePath, "resourcePath must not be null");

        try (InputStream inputStream = Thread.currentThread().getContextClassLoader().getResourceAsStream(resourcePath)) {
            if (inputStream == null) {
                throw new IllegalStateException("Recommendation demo scenarios resource not found: " + resourcePath);
            }

            Object loaded = new Yaml().load(inputStream);
            if (!(loaded instanceof Map<?, ?> root)) {
                throw new IllegalStateException("Recommendation demo scenarios must be a YAML object: " + resourcePath);
            }

            return new RecommendationDemoScenarios(scenarios(root.get("scenarios")));
        } catch (IOException exception) {
            throw new IllegalStateException("Failed to load recommendation demo scenarios: " + resourcePath, exception);
        }
    }

    private static List<RecommendationDemoScenario> scenarios(Object value) {
        if (!(value instanceof List<?> rawList)) {
            throw new IllegalStateException("Expected scenarios YAML list");
        }

        List<RecommendationDemoScenario> scenarios = new ArrayList<>();
        for (Object rawScenario : rawList) {
            if (!(rawScenario instanceof Map<?, ?> scenarioMap)) {
                throw new IllegalStateException("Scenario must be an object");
            }
            scenarios.add(new RecommendationDemoScenario(
                    stringValue(scenarioMap.get("id")),
                    stringValue(scenarioMap.get("title")),
                    stringValue(scenarioMap.get("description")),
                    stringValue(scenarioMap.get("expectedRecommendedCardId")),
                    request(scenarioMap.get("request")),
                    seedRecords(scenarioMap.get("seedRecords"))
            ));
        }
        return List.copyOf(scenarios);
    }

    private static DemoRecommendationRequest request(Object value) {
        if (!(value instanceof Map<?, ?> requestMap)) {
            throw new IllegalStateException("Scenario request must be an object");
        }
        return new DemoRecommendationRequest(
                stringValue(requestMap.get("spendingMonth")),
                longValue(requestMap.get("amount")),
                stringValue(requestMap.get("merchantName")),
                optionalStringValue(requestMap.get("merchantCategory")),
                stringSet(requestMap.get("paymentTags"))
        );
    }

    private static List<DemoSpendingRecordFixture> seedRecords(Object value) {
        if (!(value instanceof List<?> rawList)) {
            throw new IllegalStateException("Scenario seedRecords must be a list");
        }

        List<DemoSpendingRecordFixture> fixtures = new ArrayList<>();
        for (Object rawRecord : rawList) {
            if (!(rawRecord instanceof Map<?, ?> recordMap)) {
                throw new IllegalStateException("Seed record must be an object");
            }
            fixtures.add(new DemoSpendingRecordFixture(
                    stringValue(recordMap.get("cardId")),
                    longValue(recordMap.get("amount")),
                    LocalDate.parse(stringValue(recordMap.get("spentOn"))),
                    stringValue(recordMap.get("merchantName")),
                    optionalStringValue(recordMap.get("merchantCategory")),
                    stringSet(recordMap.get("paymentTags"))
            ));
        }
        return List.copyOf(fixtures);
    }

    private static Set<String> stringSet(Object value) {
        if (value == null) {
            return Set.of();
        }
        if (!(value instanceof List<?> rawList)) {
            throw new IllegalStateException("Expected YAML list but got: " + value);
        }
        Set<String> values = new LinkedHashSet<>();
        for (Object item : rawList) {
            values.add(stringValue(item));
        }
        return Set.copyOf(values);
    }

    private static String stringValue(Object value) {
        if (!(value instanceof String stringValue) || stringValue.isBlank()) {
            throw new IllegalStateException("Expected non-blank string but got: " + value);
        }
        return stringValue;
    }

    private static String optionalStringValue(Object value) {
        if (value == null) {
            return null;
        }
        return stringValue(value);
    }

    private static long longValue(Object value) {
        if (value instanceof Number number) {
            return number.longValue();
        }
        throw new IllegalStateException("Expected numeric value but got: " + value);
    }
}

File: card-api/src/main/kotlin/com/jean202/cardmizer/api/demo/RecommendationDemoScenariosLoader.kt
package com.jean202.cardmizer.api.demo

import org.yaml.snakeyaml.Yaml
import java.time.LocalDate

class RecommendationDemoScenariosLoader {

    fun loadDefault(): RecommendationDemoScenarios = load(DEFAULT_RESOURCE_PATH)

    fun load(resourcePath: String): RecommendationDemoScenarios {
        val inputStream = Thread.currentThread().contextClassLoader.getResourceAsStream(resourcePath)
            ?: throw IllegalStateException("Recommendation demo scenarios resource not found: $resourcePath")

        inputStream.use { stream ->
            val loaded = Yaml().load<Any>(stream)
            val root = loaded as? Map<*, *>
                ?: throw IllegalStateException("Recommendation demo scenarios must be a YAML object: $resourcePath")

            return RecommendationDemoScenarios(scenarios(root["scenarios"]))
        }
    }

    companion object {
        const val DEFAULT_RESOURCE_PATH = "demo/recommendation-scenarios.yml"

        private fun scenarios(value: Any?): List<RecommendationDemoScenario> {
            val rawList = value as? List<*>
                ?: throw IllegalStateException("Expected scenarios YAML list")
            return rawList.map { rawScenario ->
                val scenarioMap = rawScenario as? Map<*, *>
                    ?: throw IllegalStateException("Scenario must be an object")
                RecommendationDemoScenario(
                    id = stringValue(scenarioMap["id"]),
                    title = stringValue(scenarioMap["title"]),
                    description = stringValue(scenarioMap["description"]),
                    expectedRecommendedCardId = stringValue(scenarioMap["expectedRecommendedCardId"]),
                    request = request(scenarioMap["request"]),
                    seedRecords = seedRecords(scenarioMap["seedRecords"]),
                )
            }
        }

        private fun request(value: Any?): DemoRecommendationRequest {
            val requestMap = value as? Map<*, *>
                ?: throw IllegalStateException("Scenario request must be an object")
            return DemoRecommendationRequest(
                spendingMonth = stringValue(requestMap["spendingMonth"]),
                amount = longValue(requestMap["amount"]),
                merchantName = stringValue(requestMap["merchantName"]),
                merchantCategory = optionalStringValue(requestMap["merchantCategory"]),
                paymentTags = stringSet(requestMap["paymentTags"]),
            )
        }

        private fun seedRecords(value: Any?): List<DemoSpendingRecordFixture> {
            val rawList = value as? List<*>
                ?: throw IllegalStateException("Scenario seedRecords must be a list")
            return rawList.map { rawRecord ->
                val recordMap = rawRecord as? Map<*, *>
                    ?: throw IllegalStateException("Seed record must be an object")
                DemoSpendingRecordFixture(
                    cardId = stringValue(recordMap["cardId"]),
                    amount = longValue(recordMap["amount"]),
                    spentOn = LocalDate.parse(stringValue(recordMap["spentOn"])),
                    merchantName = stringValue(recordMap["merchantName"]),
                    merchantCategory = optionalStringValue(recordMap["merchantCategory"]),
                    paymentTags = stringSet(recordMap["paymentTags"]),
                )
            }
        }

        private fun stringSet(value: Any?): Set<String> {
            value ?: return emptySet()
            val rawList = value as? List<*>
                ?: throw IllegalStateException("Expected YAML list but got: $value")
            return rawList.map { stringValue(it) }.toSet()
        }

        private fun stringValue(value: Any?): String {
            val s = value as? String
            require(!s.isNullOrBlank()) { "Expected non-blank string but got: $value" }
            return s
        }

        private fun optionalStringValue(value: Any?): String? =
            if (value == null) null else stringValue(value)

        private fun longValue(value: Any?): Long =
            (value as? Number)?.toLong()
                ?: throw IllegalStateException("Expected numeric value but got: $value")
    }
}
